treesit.el patches

garyo opened the pull request:
- Fix ts-fold-close-all
    
    `(car fold-range)` is not a sequence, it's a symbol, at least in my
    C++ buffers. This patch converts it using `symbol-name`.

-   Check that `treesit-buffer-root-node` is defined
    
    Not sure why this is needed but I get startup errors without these.
    Perhaps an ordering issue.

-    Add `ts-` modes to alist

- Fix `ts-fold-indicators`

garyo committed:
Add ts- modes

garyo committed:
Check that `treesit-buffer-root-node` is defined

Not sure why this is needed but I get startup errors without these.
Perhaps an ordering issue.

garyo committed:
Fix ts-fold-close-all

`(car fold-range)` is not a sequence, it's a symbol, at least in my
C++ buffers. This patch converts it using `symbol-name`.

garyo committed:
Remove debug message

garyo committed:
Fix ts-fold-indicators: treesit-query-compile needs list, not vector

For simplicity I just converted the `patterns` vector to a list, but
there are probably simpler ways to do this.

garyo committed:
Handle errors from treesit-buffer-root-node

It seems `treesit-buffer-root-node` returns an error when there is no
parser for a given buffer. In my case, this happens when trying to get
doc for an elisp variable; `ts-fold--tree-sitter-trigger` runs in an
elisp-mode "*temp*" buffer, and `treesit-buffer-root-node` raises an
error.
If I just ignore the error around that call,
`ts-fold--tree-sitter-trigger` still runs `(ts-fold-mode -1)`. Seems
like the simplest thing is to just do nothing in that case.

To repro the problem, enable ts-fold mode globally and do `C-H v
emacs-version`.

AndrewSwerlick commented on ts-fold.el:
```diff
@@ -90,7 +95,9 @@
     (tsx-ts-mode . ,(ts-fold-parsers-typescript))
-    (typescript-mode . ,(ts-fold-parsers-typescript)))
+    (typescript-mode . ,(ts-fold-parsers-typescript))
+    (typescript-ts-mode . ,(ts-fold-parsers-typescript))
```
I think this is duplicate of line 96

AndrewSwerlick reviewed (changes requested):
Just one duplicate line I think.

garyo committed:
Remove duplicate parser entry

garyo committed:
Add YAML support

This is not perfect; when folding a node with children, it shows the
first character of the node, then the fold indicator, then the colon.
Ideally it would show the whole name of the key, and then the fold
indicator. Not sure how to do that.

garyo committed:
Fix yaml support

- Use a better ts-fold-range function for yaml
- Remove "block_sequence" fold-parser; don't think that's useful.
- Fix ts-fold--get-nth-child: `(treesit-node-children) is a list, not
  an array.

AndrewSwerlick commented:
@garyo thanks for this!

garyo commented:
By the way, you mentioned just switching to the new `-ts-` modes in your language table; I think you need to keep both the regular and the `-ts-` modes in your table, because some folks will keep using the standard modes and some will switch to the treesit-based ones.

AndrewSwerlick commented:
> because some folks will keep using the standard modes and some will switch to the treesit-based ones.

But if folks use the standard modes, I don't think this package will even work for them. The first line of the [home page](https://emacs-tree-sitter.github.io/) of package based tree sitter integration that the original ts-fold relies on says

> For Emacs 29+, please use the built in integration instead.

So if I'm reading that right, `emacs-tree-sitter` is going to stop working for certain versions of emacs in the near future, which will also break the original ts-fold.

If we want to make a version of ts-fold that works for multiple versions of emacs, then we'd need to support both `emacs-tree-sitter` and the built in version `treesit.el`, but I'm not sure I see the value in having something like that. It seems like it would take a lot of work to build some sort of compatibility shim that would abstract all the implementation details of each treesitter integration. And the value of that work would likely be temporary once emacs 29 has relatively wide adoption.

I'd been leading towards the model of having 2 different versions of ts-fold, one legacy compat, targeted at emacs 28 <= and then the new one targeted and 29+

garyo commented:
I think your idea is exactly right; one version of ts-fold for the native treesit and keep the old one for tree-sitter.

What I meant is that I think for the time being, some folks who have emacs30 with treesit compiled in will use native treesit for various uses (and not use tree-sitter), but for their major modes, they'll continue using (e.g.) `c++-mode` because some of the new `-ts-` modes aren't quite up to par yet. You can use native treesit minor mode along with the old major modes if I understand correctly; you just don't get treesit-based navigation and syntax highlighting. I've switched over most of mine (using `major-mode-remap-alist`) but left a few back.
